modules/ip/cli/icmp.c

```diff
@@ -0,0 +1,205 @@
+
+#include <gr_api.h>
+#include <gr_cli.h>
+#include <gr_ip4.h>
+#include <gr_net_types.h>
+
+#include <ecoli.h>
+#include <rte_ip.h>
+
+#include <errno.h>
+#include <signal.h>
+#include <sys/queue.h>
+#include <unistd.h>
+
+static const char *ICMP_DEST_UNREACHABLE[] = {
+	"net unreachable",
+	"host unreachable",
+	"protocol unreachable"
+	"port unreachable",
+	"fragmentation needed and DF set",
+	"source route failed",
+};
+
+static bool stop;
+
+static void sighandler(int) {
+	stop = true;
+}
+
+static cmd_status_t icmp_request(
+	const struct gr_api_client *c,
+	struct gr_ip4_icmp_request *req,
+	uint16_t msdelay,
+	uint16_t count,
+	bool mode_traceroute
+) {
+	cmd_status_t ret_code = CMD_SUCCESS;
+	struct gr_ip4_icmp_request_id *icmp_req;
+	struct gr_ip4_icmp_response *resp;
+	void *resp_ptr = NULL;
+	int timeout;
+	int ret;
+
+	for (int i = mode_traceroute; i < count && stop == false; i++) {
+		req->ttl = (mode_traceroute ? i : 64);
+		req->sequence_number = i;
+
+		if (gr_api_client_send_recv(c, GR_IP4_ICMP_REQUEST, sizeof(*req), req, &resp_ptr)
+		    < 0) {
+			ret_code = CMD_ERROR;
+			goto exit;
+		}
+		icmp_req = resp_ptr;
+
+		timeout = 50;
+		resp_ptr = NULL;
+		do {
+			usleep(10000);
+			ret = gr_api_client_send_recv(
+				c, GR_IP4_ICMP_GET_REPLY, sizeof(*icmp_req), icmp_req, &resp_ptr
+			);
+			resp = resp_ptr;
+			if (resp && resp->answered)
+				break;
+		} while (ret == 0 && --timeout);
+		free(icmp_req);
+
+		if (ret < 0) {
+			ret_code = CMD_ERROR;
+			goto exit;
+		}
+
+		resp = resp_ptr;
+		if (resp == NULL || timeout == 0) {
+			printf("Timeout\n");
+			errno = ETIMEDOUT;
+			ret_code = CMD_ERROR;
+		} else {
+			switch (resp->type) {
+			case 0:
+				if (mode_traceroute) {
+					printf("(%d) " IP4_ADDR_FMT " time=%.3fms\n",
+					       i,
+					       IP4_ADDR_SPLIT(&req->addr),
+					       resp->response_time / 1000.);
+					stop = true;
+
+				} else {
+					printf("icmp_seq=%d ttl=%d time=%.3fms\n",
+					       resp->sequence_number,
+					       resp->ttl,
+					       resp->response_time / 1000.);
+				}
+				break;
+			case 3:
+				printf("Destination unreachable: %s\n",
+				       ICMP_DEST_UNREACHABLE[resp->code]);
+				break;
+			case 11:
+				struct rte_ipv4_hdr *hdr = (struct rte_ipv4_hdr *)resp->data;
+				printf("(%d) " IP4_ADDR_FMT " time=%.3fms\n",
+				       hdr->time_to_live,
+				       IP4_ADDR_SPLIT(&hdr->src_addr),
+				       resp->response_time / 1000.);
+				break;
+			}
+			free(resp_ptr);
+			usleep(msdelay * 1000);
+		}
+	}
+exit:
+	return ret_code;
+}
+
+static cmd_status_t ping(const struct gr_api_client *c, const struct ec_pnode *p) {
+	struct gr_ip4_icmp_request req = {.sequence_number = 0};
+	cmd_status_t ret_code = CMD_SUCCESS;
+	uint16_t msdelay = 1000;
+	uint16_t count = 4;
+
+	if (inet_pton(AF_INET, arg_str(p, "IP"), &req.addr) != 1) {
+		errno = EINVAL;
+		return CMD_ERROR;
+	}
+
+	stop = false;
+	sighandler_t prev_handler = signal(SIGINT, sighandler);
```

Comment:
I think signal() is not that portable. And sigaction is usually preferred. Not sure it is important for grout though.

Reply:
signal is already used in other parts of grout.
Maybe we should fix all them in a separate patch ?

Reply:
Ok for now.